### PR TITLE
[WIP] Vault: Loadout-based Activity Reward System

### PR DESCRIPTION
**Pull request by Doster-d**
## About The Pull Request

This pull request adds the Vault.
Each player has their own individual Vault, where some items from the previous round are stored.
If the player hasn't sold these items, they appear in their possession from the beginning of the round.
In addition (there will be) a link to a donation system (Patreon), whose members will be able to conduct transactions over iriski more favorably than regular players.
image

## TODO LIST

- [x] Persistence System (Load & Save test included)
- [x] Vault Basics (currency, buy/sell actions, ckey-based items)
- [ ] DB&Patreon integration
- [ ] Integration test systems
- [ ] Metric Collection and Rewarding
- [ ] Vault Content
- [ ] Massive Test (if possible)

## Why It's Good For The Game

Players get rewarded for their play and stay longer with us

## Testing

The following tests were conducted:
- Buying and selling items
- Attempting to go negative on Iriski
- Loading and unloading items

## Changelog
:cl:
add: Vault Loadout-based Activity Reward System is now active. You can earn points (iriski) and buy some items and perks!
/:cl:

### CODE REVIEW

**Comment by SirRichardFrancis** [comment]
> (Patreon), whose members will be able to conduct transactions over iriski more favorably than regular players.

I thought giving patrons gameplay advantage over rest of the player base was a huge no-no, but apparently that have changed.

**Comment by Mycah142** [comment]
Fuck the poor.

**Comment by Doster-d** [comment]
> I thought giving patrons gameplay advantage over rest of the player base was a huge no-no, but apparently that have changed.

But it's not gameplay advantage (in my opinion), for example you have item that costs 5 iriski, if you're patron you can sell it for 7 iriski.

However the amount of discount and the discount itself is subject to change.

**Comment by Benblu** [comment]
Remove the patron discount and maybe this will actually be palatable.

**Comment by thevandie** [comment]
there's like 4 problems (with solutions) to this
1) People are gonna carry over crazy shit like dallases or bloodred voids from previous rounds, and the roundstart balance is gonna be FUCKED
2) People are gonna break the game by carrying over one-time items like faction artifacts
3) Carrying antag exclusive items over from previous rounds would be FUCKED, cause it'd mean you literally have zero idea who a traitor is and who isn't, where-as currently we at least have a small idea (it's still fucked, but at least you can realistically tell someone's an antag if they have an uplink).
4) If you remove these specific items it'll delete some of the purpose to this whole idea

SO, first suggestion would be to add a banned items list. Stop some items from being carried over at all.
Second suggestion, to keep the incentive on still carrying these items over the round, maybe just turn them straight into cash for the next round, or iriska coin I guess
For example, if you hold a Dallas over, it forces you to sell it. However, it gives you a massive payout cause it's a super rare item. I suppose you could do a similar thing with completing faction and antag objectives in general, just give people money for it, but that doesn't sound like it's in the scope of the PR so I won't expand further. You should also probably equate antag items to their TC cost, so people don't just buy (for example) loads of BSDM units and absolutely game the system by getting thousands of iriska coins in one round. So instead, 1 TC should be worth the same amount no matter what it's spent on. That should also be accounted for if it's still in the uplink, btw.

TLDR, make a ban list on items you can carry over, but give uniquely massive payouts for grabbing them
like faction artifacts, any antag items, and rare combat items, and make everything that uses TC worth the same amount as the individual TeleCrystals.

**Comment by thevandie** [comment]
As a personal note, I actually really like this idea
I've been playing the recent Blue Colony Reborn server, and the ability to carry over items from previous rounds is an interesting concept that I enjoy
Although obviously it can't be taken to that server's extreme, since you can't just store stuff in a lot. Still, being able to carry over the same gun through your rounds is an amazing idea and I love it.

**Comment by null-Aurelian** [comment]
well if we can force sell certain items, it might be an idea to add faction artifacts to the force-sold list, sort of like a running king of the hill for personal benefit. It incentivizes theft of artifacts without destruction by assorted randoms.

**Comment by thevandie** [comment]
> well if we can force sell certain items, it might be an idea to add faction artifacts to the force-sold list, sort of like a running king of the hill for personal benefit. It incentivizes theft of artifacts without destruction by assorted randoms.

That's literally exactly what I said

**Comment by Mycah142** [comment]
Alternative: We use the Spawn_Blacklist as a reference of items that shouldn't be carried over, and instead of letting them be carried over, maybe they can be exchanged for coins without the player having a say? This might help avoid shenanigans. I dunno if anyone said any of this before, I saw alot of words and my eyes glazed over.

**Comment by tristain01** [comment]
Perosonally, i think this is going to break the game, be crap in the sense you cant store anything fun or some where inbetween.

**Comment by null-Aurelian** [comment]
TBH I thought the Iriski were going to be tied to round scores and give funny/painful things like "start with a corgi" or "start with a hologun" or "sanity is capped to 0"

**Comment by Doster-d** [comment]
In the "Metric Collection and Rewarding" you will receive by calculation of your actions in the end (probably round stats) 3-4 max "item points", that you can currently spend on saving 3-4 items or perks.

There will be a blacklist of items, but the system in design allows you to store "fun or anything in between".

In the next round, if you don't sell this things, you will spawn with them.

The selling price will be much lower (½ of price) than price of buying (the price of item itself)


YOU CAN'T SELECT ITEMS, THEY WILL BE SELECTED BY YOUR ACTIONS (RNG ALARM)


Arguably you can spare iriski to get more items in your Vault at round end.


Donation p2w system:

The discount is subject to change.

You can't buy iriski from real or in-game bucks. You get them by playing in game and do things that game thinks you should do (completing antag goals, your division goals)

If you're max tier you (will) have 3-4 iriski bonus, e.g., item price is 6 for everyone, but you can sell it for 6+3-4 iriski.

Probably it will just changed to get iriski bonus at the round end iriski giveaway, but at this stage of development I made up this system to test things.

**Comment by Firefox13** [comment]
> Probably it will just changed to get iriski bonus at the round end iriski giveaway, but at this stage of development I made up this system to test things.

This is a more welcome solution rather than selling price increase.

> In the next round, if you don't sell this things, you will spawn with them.

Ability to store them is also intended.

> YOU CAN'T SELECT ITEMS, THEY WILL BE SELECTED BY YOUR ACTIONS (RNG ALARM)

I probably should've mentioned this earlier, but it shouldn't be the items that you physically carry out.
Internally, there should be several separate lists of items for iriska vault system - perks, oddities (weak ones), guns (unlikely, very OP), fun shit that shouldn't go in loadout (paintjobs for weapons for patrons) - and use them for rewarding and internal shop.
This solves the "people will hoard ten dallases round-end or get completely nude to get that one item" issue. I really don't want that to happen.
This ALSO solves the "Ided and didn't get any items" issue. 
Of course, if you died mid-round the reward will be lower but we don't want players just stop having conflict and do nothing for three hours straight to get some iriskas. How to solve this:
1. Rewards must not be very valuable. No "extremely rare Dallas" or "+50 sanity perk".
2. Observers simply do not get anything. To be eligible for a reward you should've played a certain portion of the round, as a human/blitz/robot/borer/any other antag (not mouse) and *be* there during roundend. Not necessarily alive. If round lasted for 3 hours and your character lasted for 30 minutes you should be eligible for rewards (numbers still need thinking)
@Doster-d

**Comment by Doster-d** [comment]
Okay discount will be removed closer to "Metric Collection and Rewarding" and will be replaced iriski bonus at round end.
I will be adding ability to store items indefintly asap, as well as changing some internal stuff.

**Comment by TheLeopold** [comment]
> it shouldn't be the items that you physically carry out

Then why do you call it "Vault" and not, i dunno, "Loadout 2.0"?

**Comment by Firefox13** [comment]
> Then why do you call it "Vault" and not, i dunno, "Loadout 2.0"?

that's how it's called in design doc
it also sounds really cool

**Comment by Mycah142** [comment]
I'm going to make a thread on the discord somewhere so this git doesn't get super crowded.

**Comment by Mycah142** [comment]
https://discord.com/channels/255035529085583360/1151870542970896384/1151870546993233990

Questions are being answered, and discussion can happen here.

**Comment by TheShown911** [comment]
Cyberpunk enjoyers when the ritch get a gameplay advantage

**Comment by Alliostra** [comment]
battlepass_preorderbonus_horsearmor_lootboxes_weeklyloginbonus_gacha

**Comment by thevandie** [comment]
pretty sure this hasn't been looked at cause it's still at 7/8 tasks and looks unfinished (context or not)
if someone is reading the github history from discord, get on this shiet

**Comment by Doster-d** [comment]
Description updated.
FireFox was informed about status of this PR 10/06/23

**Comment by thevandie** [comment]
Based 👍

**Comment by hyperioo** [comment]
Smol conflict for tgui, quick rebase and rebuilding tgui should fix it

**Comment by Doster-d** [comment]
It should be done

**Comment by Kurgis** [comment]
Does the Patron status violate anything within BYOND's terms of service/EULA, and our ability to be on the hub?

**Comment by thevandie** [comment]
It's apparently done, don't forget,  big people

**Comment by thevandie** [comment]
It's apparently done, Firefox,  hyperio and/or shown, please make sure it gets looked at Friday
I beg of thee

**Comment by Firefox13** [comment]
The system needs *content*. Think of its like R&D but without any tech trees coded in. Adding it in without rewards added in and without (atleast somewhat) balanced rewards is pointless.

**Comment by Doster-d** [comment]
Well. It's done. It should work.

**Comment by Doster-d** [comment]
If there is any brave soul who would like to play and test it, I can assure you I will provide you support in this hard task
